### PR TITLE
CMake IREE_ARCH variable, a canonicalized CMAKE_SYSTEM_PROCESSOR

### PR DESCRIPTION
**Pull request by bjacob**
As suggested by @GMNGeoffrey on #12674.

### CODE REVIEW

File: build_tools/cmake/iree_macros.cmake [hunk 1]
```diff
@@ -19,6 +19,75 @@ else()
...
+# non-trivial: it usually is CMAKE_SYSTEM_PROCESSOR, but on some platforms, we
+# have to read other variables instead.
+if(CMAKE_OSX_ARCHITECTURES)
+  list(LENGTH CMAKE_OSX_ARCHITECTURES NUM_ARCHES)
+  if(NOT ${NUM_ARCHES} EQUAL 1)
+    # Leaving _IREE_UNNORMALIZED_ARCH empty disables arch code paths. We will
+    # issue a performance warning about that below.
+  endif()
+  # Only one arch in CMAKE_OSX_ARCHITECTURES, use that.
+  set(_IREE_UNNORMALIZED_ARCH "${CMAKE_OSX_ARCHITECTURES}")
```
Comment (GMNGeoffrey) [inline build_tools/cmake/iree_macros.cmake]: This if block is empty?

```suggestion
  if(${NUM_ARCHES} EQUAL 1)
    # Only one arch in CMAKE_OSX_ARCHITECTURES, use that.
    set(_IREE_UNNORMALIZED_ARCH "${CMAKE_OSX_ARCHITECTURES}")
  endif()
  # Leaving _IREE_UNNORMALIZED_ARCH empty disables arch code paths. We will
  # issue a performance warning about that below.
```

File: build_tools/cmake/iree_run_module_test.cmake [hunk 2]
```diff
@@ -5,10 +5,13 @@
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 # Utility function to return the platform name in crosscompile.
+# TODO: stop leaking CMAKE_SYSTEM_PROCESSOR values. Try to harmonize around
```
Comment (GMNGeoffrey) [inline build_tools/cmake/iree_run_module_test.cmake]: Can you link an issue so this doesn't get lost? I guess the real issue is the usage of this in `iree_benchmark_suite_module_test` and the generated CMake that uses that? @pzread should be able to help. Should just be a matter of updating the names of the platforms in `XFAIL_PLATFORMS` and `UNSUPPORTED_PLATFORMS`, I think

File: runtime/src/iree/builtins/ukernel/arch/CMakeLists.txt [hunk 3]
```diff
@@ -12,19 +12,8 @@ else()
   set(IREE_UK_ENABLE_ARCH_SPECIFIC_CODE TRUE)
 endif()
 
-# This block is borrowed from boringssl's CMake code here:
-# https://boringssl.googlesource.com/boringssl/+/c5f0e58e653d2d9afa8facc090ce09f8aaa3fa0d/CMakeLists.txt#43
```
Comment (GMNGeoffrey) [inline runtime/src/iree/builtins/ukernel/arch/CMakeLists.txt:16]: Maybe preserve this heritage in the new function?

**Review by GMNGeoffrey — COMMENTED** [review]
Curious about the decision to use "arm64" rather than "aarch64". I thought the latter was "more correct". Should we update target_platforms.h to use that rather than canonicalizing this direction?

**Comment by bjacob** [comment]
> Curious about the decision to use "arm64" rather than "aarch64". I thought the latter was "more correct". Should we update target_platforms.h to use that rather than canonicalizing this direction?

I am merely following the existing IREE-wide convention here, we name it `arm_64` consistently in the `IREE_ARCH` token in `target_platform.h` , in other preprocessor suffixes as in `IREE_ARCH_ARM_64`, and in filename suffixes (so harmonizing the other way would involve renaming a number of files).

Honestly there is just no consistency as to what is the correct name. The term `arm64` is widely used in the Linux kernel, in Android, and in MSVC. In ARM's own architecture documentation, it is called either `AArch64` or `A64`, so there isn't perfect consistency even there. Also, consider that 'Arch' stands for 'architecture', it is actually redundant here and not conveying useful information. The only useful letter in `AArch64` is thus the `A`, which stands for `ARM`. In that sense, `arm64` is a better use of letters, it avoids trademarking the letter `A` :-)

IREE's use of a `_` to separate the `64` makes us a little more non-standard on each architecture (semi-standard `arm64` -> `arm_64`) in exchange for buying us consistency across architectures. It grew on me: now in IREE contexts I always know how each architecture is named. Otherwise I'd have to check everytime if it's `riscv64`, `riscv-64` or `riscv_64`.

**Comment by bjacob** [comment]
@ScottTodd it's useful that you take a quick look at this PR, search for `EMSCRIPTEN`.

**Comment by GMNGeoffrey** [comment]
> The only useful letter in `AArch64` is thus the `A`, which stands for `ARM`. In that sense, `arm64` is a better use of letters, it avoids trademarking the letter `A` :-)

Ha good point! :-) As long as we're not using the "wrong" thing then I agree that arm is clearer. I also never know how to pronounce "AArch"

**Review by GMNGeoffrey — APPROVED** [review]

**Pull request merged**